### ISSUE TITLE
fails to run freeipa-container on Fedora31 with podman

### ISSUE THREAD

**Issue by waulee** [freeipa/freeipa-container#317]
A am unable to start container built with command: 
```
[root@cntr01 freeipa-container]# podman build -t localhost/freeipa-container .

[root@cntr01 freeipa-container]# docker run -ti --name freeipa-master -h ipa.example.test --sysctl net.ipv6.conf.all.disable_ipv6=0 --tmpfs /run --tmpfs /tmp -v /dev/urandom:/dev/random:ro -v /sys/fs/cgroup:/sys/fs/cgroup:ro -v /tmp/freeipa-test-351458/data:/data:Z -e PASSWORD=Secret123 localhost/freeipa-container exit-on-finished -U -r EXAMPLE.TEST --setup-dns --no-forwarders --no-ntp
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
systemd v243.8-1.fc31 running in system mode. (+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=unified)
Detected virtualization container-other.
Detected architecture x86-64.
Set hostname to <ipa.example.test>.
Initializing machine ID from random generator.
Failed to create /init.scope control group: Read-only file system
Failed to allocate manager object: Read-only file system
[!!!!!!] Failed to allocate manager object.
Exiting PID 1...
```

I can run a the fedora31 container with systemd succecfully:
```
[root@cntr01 freeipa-container]# podman run --rm -h test1.test.local -ti --name systemd registry.fedoraproject.org/fedora:31 /usr/sbin/init
systemd v243.7-1.fc31 running in system mode. (+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=unified)
Detected virtualization container-other.
Detected architecture x86-64.

Welcome to Fedora 31 (Container Image)!

Set hostname to <test1.test.local>.
Initializing machine ID from random generator.
[  OK  ] Started Dispatch Password Requests to Console Directory Watch.
[  OK  ] Started Forward Password Requests to Wall Directory Watch.
[  OK  ] Reached target Local File Systems.
[  OK  ] Reached target Paths.
[  OK  ] Reached target Remote File Systems.
[  OK  ] Reached target Slices.
[  OK  ] Reached target Swap.
[  OK  ] Listening on Process Core Dump Socket.
[  OK  ] Listening on initctl Compatibility Named Pipe.
[  OK  ] Listening on Journal Socket (/dev/log).
[  OK  ] Listening on Journal Socket.
         Starting Rebuild Dynamic Linker Cache...
         Starting Journal Service...
         Starting Create System Users...
[  OK  ] Started Create System Users.
[  OK  ] Started Rebuild Dynamic Linker Cache.
[  OK  ] Started Journal Service.
         Starting Flush Journal to Persistent Storage...
[  OK  ] Started Flush Journal to Persistent Storage.
         Starting Create Volatile Files and Directories...
[  OK  ] Started Create Volatile Files and Directories.
         Starting Rebuild Journal Catalog...
         Starting Update UTMP about System Boot/Shutdown...
[  OK  ] Started Update UTMP about System Boot/Shutdown.
[  OK  ] Started Rebuild Journal Catalog.
         Starting Update is Completed...
[  OK  ] Started Update is Completed.
[  OK  ] Reached target System Initialization.
[  OK  ] Started Daily Cleanup of Temporary Directories.
[  OK  ] Reached target Timers.
[  OK  ] Listening on D-Bus System Message Bus Socket.
[  OK  ] Reached target Sockets.
[  OK  ] Reached target Basic System.
         Starting Permit User Sessions...
[  OK  ] Started Permit User Sessions.
[  OK  ] Reached target Multi-User System.
         Starting Update UTMP about System Runlevel Changes...
[  OK  ] Started Update UTMP about System Runlevel Changes.
```

**Comment by adelton** [comment]
You say it fails with podman but then show failed `docker run`. Is that just a typo? If yes and things work with docker and not with podman and you have the recent podman 1.9.0, you might be hitting https://bugzilla.redhat.com/show_bug.cgi?id=1827315 which I happened to report yesterday. IOW, try to turn off `systemd.unified_cgroup_hierarchy=0`.

**Comment by waulee** [comment]
No I have tried that way too, but lastly I have installed podman-docker with dnf install podman-docker.
I will try it,
Thank You

**Comment by adelton** [comment]
Which podman version is that exactly?

**Comment by waulee** [comment]
The latest so it is 1.9.0.
But I am trying it with the root user it is not rootless a conatiner!

[root@cntr01 freeipa-container]# podman info
host:
  arch: amd64
  buildahVersion: 1.14.8
  cgroupVersion: v2
  conmon:
    package: conmon-2.0.15-1.fc31.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.15, commit: 4152e6044da92e0c5f246e5adf14c85f41443759'
  cpus: 4
  distribution:
    distribution: fedora
    version: "31"
  eventLogger: journald
  hostname: cntr01.local
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.5.17-200.fc31.x86_64
  memFree: 2943328256
  memTotal: 8341995520
  ociRuntime:
    name: crun
    package: crun-0.13-2.fc31.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 0.13
      commit: e79e4de4ac16da0ce48777afb72c6241de870525
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  rootless: false
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 0
  swapTotal: 0
  uptime: 75h 29m 9.69s (Approximately 3.12 days)
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - registry.centos.org
  - docker.io
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 11
    paused: 0
    running: 0
    stopped: 11
  graphDriverName: overlay
  graphOptions:
    overlay.mountopt: nodev,metacopy=on
  graphRoot: /var/lib/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "true"
  imageStore:
    number: 108
  runRoot: /var/run/containers/storage
  volumePath: /var/lib/containers/storage/volumes

**Comment by waulee** [comment]
[root@cntr01 freeipa-container]# podman -v
podman version 1.9.0

Sorry I missed it.

**Comment by adelton** [comment]
Can you try removing the `-v /sys/fs/cgroup:/sys/fs/cgroup:ro` from the `podman run` parameters?

**Comment by waulee** [comment]
The same result:
[root@cntr01 freeipa-container]# podman run -ti --name freeipa-master -h ipa.example.test --sysctl net.ipv6.conf.all.disable_ipv6=0 --tmpfs /run --tmpfs /tmp -v /dev/urandom:/dev/random:ro  -v /tmp/freeipa-test-351458/data:/data:Z  -e PASSWORD=Secret123 localhost/freeipa-container exit-on-finished -U -r EXAMPLE.TEST --setup-dns --no-forwarders --no-ntp
systemd v243.8-1.fc31 running in system mode. (+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=unified)
Detected virtualization container-other.
Detected architecture x86-64.
Set hostname to <ipa.example.test>.
Initializing machine ID from random generator.
Failed to create /init.scope control group: Read-only file system
Failed to allocate manager object: Read-only file system
[!!!!!!] Failed to allocate manager object.
Exiting PID 1...

**Comment by waulee** [comment]
I am unable to start the container unless --privileged used.
But I am aware of that this is no a good solution but a work-around.

**Comment by waulee** [comment]
Now that the container fired up and the freeipa installation begun, but at some point it failed just like it did in issue #309. https://github.com/freeipa/freeipa-container/issues/309
My environment is different I used Fedora31 as host-os not MAC.

**Comment by waulee** [comment]
On my other VM I use centos-8 as host-os. The containr is also centos-8. The installation of freeipa in container went flawless.

**Comment by waulee** [comment]
If I downgrade to podman 1.6.2 and try to run the container I got the same error with a little bit different error message:

[root@cntr01 ~]# podman  run -ti --name freeipa-master -h ipa.example.test --sysctl net.ipv6.conf.all.disable_ipv6=0 --tmpfs /run --tmpfs /tmp -v /dev/urandom:/dev/random:ro -v /sys/fs/cgroup:/sys/fs/cgroup:ro -v /tmp/freeipa-test-351458/data:/data:Z -e PASSWORD=Secret123 localhost/freeipa-container exit-on-finished -U -r EXAMPLE.TEST --setup-dns --no-forwarders --no-ntp
systemd v243.8-1.fc31 running in system mode. (+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=unified)
Detected virtualization container-other.
Detected architecture x86-64.
Set hostname to <ipa.example.test>.
Initializing machine ID from random generator.
Failed to create /machine.slice/libpod-825f104374fdb7e10615e616ce434abaa5dddeb72b72829b9f95be22ec2e74e0.scope/init.scope control group: Read-only file system
Failed to allocate manager object: Read-only file system
[!!!!!!] Failed to allocate manager object.
Exiting PID 1...

As far as I am understand shouldn't the systemd (/usr/sbin/init) be the first process that is executed in the container? I see that You do some initialization with a shell script befor executing the real init.

**Comment by adelton** [comment]
Do you have the `container_manage_cgroup` SELinux boolean set?
Yes, we do initialization of the `/data` volume before running the systemd.

**Comment by waulee** [comment]
Yes, it is set and selinux is enforcing:
[root@cntr01 ~]# getsebool container_manage_cgroup
container_manage_cgroup --> on

[root@cntr01 ~]# getenforce 
Enforcing

And the audit log is clear, nothing logged as denied.

Now I turned selinux to permissive (setenforce 0), but running the freeipa-container without --privileged podman option is not possible. I got the same error as before.

Just for curiosity, shouldn't be it better way, to initialize /data with a systemd service (one-shot) at container starup before ipa services starts up and let systemd handle everything in the container?
I see all of the target systems have systemd atleast the recent one-s. (I don't remember exactly if fedora 24 had one or not, it was a while I used it.)

**Comment by adelton** [comment]
To debug this further, you might want to try to change the Dockerfile as
```
diff --git a/Dockerfile.fedora-31 b/Dockerfile.fedora-31
index 7893595..252e7f4 100644
--- a/Dockerfile.fedora-31
+++ b/Dockerfile.fedora-31
@@ -53,7 +53,7 @@ RUN mv /usr/bin/nisdomainname /usr/bin/nisdomainname.orig
 ADD hostnamectl-wrapper /usr/bin/nisdomainname
 
 # test-addon: VOLUME [ "/var/log/journal" ]
-## # test: systemd-container-ipa-server-install.sh
+# test: systemd-container-ipa-server-install.sh
 
 # Move configuration and data to data volume
 COPY patches/ipa-data-fedora-30.patch /root
```
and then run
`tests/run-partial-tests.sh Dockerfile.fedora-31`. That will run some rudimentary tests with plain systemd, and then run the `ipa-server-install` without the `/data` logic applied.

As for the `/data` initialization, since it contains things like `/etc/systemd/system` or `/var/log/journal`, we need to setup its content before we start systemd.

**Comment by adelton** [comment]
Were you able to do some additional testing?

**Comment by waulee** [comment]
Hi! Sorry for the delay.
I have just applied the patch, and the requested tests/run-partial-tests.sh Dockerfile.fedora-31starts a container, the installation process begins.
But the dynamicaly attached voluem for journal does not conatin any journal as it is probably sent to only /run/log/journal.
After the free-ipa installation process finished. A new container is created and that container didn't start up.

**Comment by Tiboris** [comment]
I believe this is a problem with cgroups V2 
https://bugzilla.redhat.com/show_bug.cgi?id=1762972
https://bugzilla.redhat.com/show_bug.cgi?id=1834814

**Comment by waulee** [comment]
Well thank You pointing this out. 
I have just tested the referenced problem too. The ubi8 image complained that it was not booted as systemd with entrypoint. Striping down the tmpfs and the volume options I got this error:

# podman run -t -d --name test   localhost/test_init_wrapper && sleep 10 && podman exec test systemctl status ; podman logs test
dbecc24c3ae81be5453573573c08c2f9e92907c3b294ca27b1588ab485ca9cf2
**_System has not been booted with systemd as init system (PID 1). Can't operate._**
Failed to connect to bus: Host is down
Error: exec session exited with non-zero exit code 1: OCI runtime error
Failed to mount tmpfs at /run: Operation not permitted
[!!!!!!] Failed to mount API filesystems, freezing.
Freezing execution.

I dont realy know if this is a bug or a security feature of the systemd in the conatiner environment. There are some considerations about system liability and auditability if one takes these into account. Perhaps it is a way how cgroups2 and systemd enforces this.
As I am understand there are other means of initializing directories from template and symlinks before real application starts. ( ex. /etc/tmpfiles.d/ and other systemd services hooked before the application starts)
Dont get me wrong it is perfectly OK that a process makes preps for the application. These are perfect for many kind of applications. I just feel that when it's about security one should be more precautious.

**Comment by adelton** [comment]
You say
> the requested tests/run-partial-tests.sh Dockerfile.fedora-31 starts a container, the installation process begins.

So that should give you an example of running the FreeIPA container on your machine, right?

The journal is in the data volume (`/data/var/log/journal`), not in the tmpfs volume `/var/log/journal`.

You say
>  After the free-ipa installation process finished. A new container is created and that container didn't start up.

Can you show the exact `tests/run-partial-tests.sh Dockerfile.fedora-31` output to show what you are getting?

BTW, please use \``` ... \``` to quote your output to keep it verbatim.

**Comment by adelton** [comment]
If you've seen the `ipa-server-install` process to start when you run `tests/run-partial-tests.sh Dockerfile.fedora-31`, I'd say you are past the hurdle of running systemd in the container. Just use the same `docker run` (== `podman run`) parameters that the test script uses.

**Comment by waulee** [comment]
Unfortunatly I am unable to capture the whole output of the run-tarial-test.sh. If I am trying to pipe to tee I got permission denied error from tee opening /dev/stderr at some point where patches are applied to python scripts:
Be sure to back up the CA certificates stored in /root/cacert.p12
These files are required to create replicas. The password for these
files is the Directory Manager password
tail: freeipa-container-test-run-partial-tests.out: file truncated
-rw-------. 1 root root 4015836 May 12 21:14 /var/log/ipaserver-install.log
Error: exec session exited with non-zero exit code 1: OCI runtime error
-rw-r-----+ 1 root systemd-journal 8388608 May 12 21:14 /run/log/journal/d21d9e75e00f4a28b1518cf60aedb5e8/system.journal
**_OK tests/systemd-container-ipa-server-install.sh._**
STEP 1: FROM local/freeipa-server-test:fedora-31
STEP 2: COPY patches/ipa-data-fedora-30.patch /root
--> Using cache 3767157d7a7add130eb80f414dcb6c890e53e6c2f4c64ec70a46cd4cfe7bda8d
STEP 3: RUN set -o pipefail ; patch --verbose -p0 --fuzz=0 < /root/ipa-data-fedora-30.patch | tee /dev/stderr | sed -n 's/^patching file //;T;/\.py$/p' | xargs python3 -m compileall
**tee: /dev/stderr: Permission denied**
Compiling '/usr/lib/python3.7/site-packages/ipaplatform/base/paths.py'...
Compiling '/usr/lib/python3.7/site-packages/ipaplatform/redhat/paths.py'...
Compiling '/usr/lib/python3.7/site-packages/ipaplatform/fedora/paths.py'...
Compiling '/usr/lib/python3.7/site-packages/ipaserver/install/adtrustinstance.py'...
Compiling '/usr/lib64/python3.7/shutil.py'...
Error: error building at STEP "RUN set -o pipefail ; patch --verbose -p0 --fuzz=0 < /root/ipa-data-fedora-30.patch | tee /dev/stderr | sed -n 's/^patching file //;T;/\.py$/p' | xargs python3 -m compileall": error while running runtime: exit status 1

None the less I am trying to capture is from my terminal:
the test runs smoothly for the ipa-server-install en the ipa-client-install part.
That container starts up and this install test succeed as You can se it form above.

If I don't run the test through tee's pipe I got further but still the test failes:

....

STEP 11: RUN /usr/local/bin/prepare-volume-template /etc/volume-tmp-list /tmp
'/root/.cache' -> '/tmp/root/.cache'
'/root/.ipa_cache' -> '/tmp/root/.ipa_cache'
'/root/.ipa/log' -> '/tmp/root/.ipa/log'
'/var/lib/certmonger' -> '/tmp/var/lib/certmonger'
'/var/lib/gssproxy' -> '/tmp/var/lib/gssproxy'
'/var/lib/chrony' -> '/tmp/var/lib/chrony'
'/var/tmp' -> '/tmp/var/tmp'
--> 7e026d4b9c8
STEP 12: RUN /usr/local/bin/extract-rpm-upgrade-scriptlets
--> d4cc6785e7e
STEP 13: RUN echo 2.0 > /etc/volume-version
--> d9b8e7b98d0
STEP 14: VOLUME [ "/tmp", "/run", "/data", "/var/log/journal" ]
--> 17cb855ae5c
STEP 15: COPY init-data-minimal /usr/local/sbin/init
--> ee817077602
STEP 16: ENTRYPOINT [ "/usr/local/sbin/init" ]
STEP 17: COMMIT local/freeipa-server-test:fedora-31
--> 65718559aa3
65718559aa3acc1a69917843c5481beb3b0ca9b9f713b593a2d338a5e3d33ed8
STEP 1: FROM local/freeipa-server-test:fedora-31
STEP 2: COMMIT local/freeipa-server-test-addons:fedora-31
--> 65718559aa3
65718559aa3acc1a69917843c5481beb3b0ca9b9f713b593a2d338a5e3d33ed8
5ffc19b6037413ed282b901929daf05bd545a9f53acb2716bfaec7abfabeacdb
Executing tests/systemd-container-ipa-server-install-data.sh freeipa-server-container-fedora-31 docker-diff-minimal-fedora-23.out
Error: can only create exec sessions on running containers: container state improper
Error: can only create exec sessions on running containers: container state improper

 And that container is not runable:

[root@cntr01 freeipa-container]# podman run -ti -h ipa.example.test --tmpfs /run --tmpfs /tmp -v /sys/fs/cgroup:/sys/fs/cgroup:ro --sysctl net.ipv6.conf.all.disable_ipv6=0 local/freeipa-server-test-addons:fedora-31
systemd v243.8-1.fc31 running in system mode. (+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=unified)
Detected virtualization container-other.
Detected architecture x86-64.

Welcome to Fedora 31 (Container Image)!

Set hostname to <ipa.example.test>.
Initializing machine ID from random generator.
Failed to create /init.scope control group: Read-only file system
Failed to allocate manager object: Read-only file system
[!!!!!!] Failed to allocate manager object.
Exiting PID 1...

Unless I add the cgroupns flag stating it is V1 mode with --cgroupns host and -v /sys/fs/cgroup:/sys/fs/cgroup:rw (read-write mode) I am not able to start the container at all. I am not 100% sure that this is the correct way:

[root@cntr01 freeipa-container]# podman run -ti -h ipa.example.test --tmpfs /run --tmpfs /tmp -v **/sys/fs/cgroup:/sys/fs/cgroup:rw** **--cgroupns host** -v journal:/var/log/journal/ -v 0e25dea770cd423761b9d90c03c439e0a5290cb3b0c0a6c2b027da2f827a58d5:/data --sysctl net.ipv6.conf.all.disable_ipv6=0 local/freeipa-server-test-addons:fedora-31
systemd v243.8-1.fc31 running in system mode. (+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=unified)
Detected virtualization container-other.
Detected architecture x86-64.

Welcome to Fedora 31 (Container Image)!

Set hostname to <ipa.example.test>.
Initializing machine ID from random generator.
[  OK  ] Listening on Journal Socket (/dev/log).
[  OK  ] Listening on Journal Socket.
         Starting Journal Service...
[  OK  ] Started Journal Service.
         Starting Create Volatile Files and Directories...
[  OK  ] Started Create Volatile Files and Directories.
[  OK  ] Reached target System Initialization.
[  OK  ] Listening on D-Bus System Message Bus Socket.
         Starting D-Bus System Message Bus...
[  OK  ] Started D-Bus System Message Bus.
[  OK  ] Reached target Minimal target for containerized FreeIPA server.

**Comment by adelton** [comment]
That should not be needed, podman should detect based on the `ENTRYPOINT` basename that it ought to prepare the environment for systemd.

Your best bet is problably to upgrade to Fedora 32, to the latest podman, and to CGroups V2.

**Comment by waulee** [comment]
Yesterday I had upgraded at first: podman to Version : 1.9.2 Release : 1.fc31: same failure. 
Then upgarded the whole OS to Fedora32 as You sugested: podman Version : 1.9.1, Release : 1.fc32: again the same failure. 
Today I have tested on this F32 system the containter images also with F31, F32 and CentOS-8 but I got the same failure.

Never mind I will back off to a virtual machine instead of the containerized for now. Once I will have some spare time I probably will try to restructure Your conatiner environment to be able to run in a CGroups V2 environment too.
Thanks for Your time.
